### PR TITLE
Fix tests to run as process

### LINKED ISSUE #160: CLI - Fix test env
We have the ability to run our tests as a process where we can run the CLI as a customer would, against the CLI executable where python is not required on the machine

We have started the work but it does not fully work yet.

### PR DESCRIPTION
**Pull request by sigalsax**
### What does this PR do?
This PR adds the ability to run our tests as a process. Tests are run using the packed CLI exec, without requiring python on the machine and running against an CLI exec. 

This PR also adds the supporting documentation

To run follow the docs supplied in the PR `./integrations_tests_runner -i test_with_process -u https://conjur-server -a someaccount -l somelogin -p Myp@ssw0rd1! -f /test -c /dist/conjur --invoke_cli_as_process`

### What ticket does this PR close?
Resolves #160 

This PR is a continuation of https://github.com/cyberark/conjur-api-python3/pull/144 and https://github.com/cyberark/conjur-api-python3/pull/129

#### Change log
- [ ] The CHANGELOG has been updated, or
- [x] This PR does not include user-facing changes and doesn't require a CHANGELOG update

#### Test coverage
- [ ] This PR includes new unit and integration tests to go with the code changes, or
- [x] The changes in this PR do not require tests

#### Documentation
- [x] Docs (e.g. `README`s) were updated in this PR, and/or there is a follow-on issue to update docs, or
- [ ] This PR does not require updating any documentation

### CODE REVIEW

File: test/util/test_helpers.py [hunk 1]
```diff
@@ -8,6 +9,18 @@ def remove_file(file_path):
...
         os.remove(file_path)
 
+def run_func_with_timeout(timeout, func,*args):
+    # important! this must raise an exception for the interrupt to work with blocking functions
+    def interrupted(signum, frame):
+        raise RuntimeError("interrupted")
+    signal.signal(signal.SIGALRM, interrupted)
+    # set alarm
+    signal.alarm(timeout)
+    s = func(*args)
```
Comment (tzheleznyak) [inline test/util/test_helpers.py:19]: give meaninful name instead of s

File: test/util/test_runners/integration_test_case.py [hunk 2]
```diff
@@ -65,25 +66,70 @@ def invoke_cli_as_code(test_runner, *args, exit_code=0):
...
+    2) interactive_input - raise a prompt withe a timeout
+    to initiate interaction with "patch" and retrieve the
+    desired value. note that the timeout is important in
+    case no input will come from "patch".
     @param test_runner:
     @param args: the cli args input
     @param exit_code:
     @return: the cli string response
     """
+    MAX_INTERACTIONS_ALLOWED = 5
```
Comment (tzheleznyak) [inline test/util/test_runners/integration_test_case.py:95]: Move const outside of the class to the top side of the file

File: conjur/client.py [hunk 3]
```diff
@@ -30,7 +30,7 @@ class ConfigException(Exception):
 
     This class is used to wrap a regular exception with a more-descriptive class name
 
-    *************** DEVELOPER NOTE ***************
+    `*************** DEVELOPER NOTE ***************`
```
Comment (tzheleznyak) [inline conjur/client.py:33]: why this change

Comment (sigalsax) [reply conjur/client.py:33]: don't know why this happened. good catch

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -18,45 +18,163 @@ any additional information for contributing code to this project.
...
-### Egg format
+### Static/portable CLI binary
 
 ```
-$ ./bin/build
+$ ./bin/build_binary
 ```
 
-### Static/portable CLI binary
+### Egg format
```
Comment (orenbm) [inline CONTRIBUTING.md:27]: what's `Egg format`?

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -18,45 +18,163 @@ any additional information for contributing code to this project.
...
 ```
 
-- Enable your terminal to use those files with this command:
+Windows:
+
+```
+py -m venv venv
+```
+
+2. Enable your terminal to use those files with this command. Note that this command will need to run each time you want to return to your virtual environment.
```
Comment (orenbm) [inline CONTRIBUTING.md:51]: i think it's best to use only `1` for lesser burden in adding future steps

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -18,45 +18,163 @@ any additional information for contributing code to this project.
...
-$ ./bin/build_binary
+$ ./bin/build
 ```
 
 ## Development
 
-- Create a directory that will hold all the virtualenv packages and files:
+To setup a development environment follow the instructions in this section. Once you have done so, you will be able to see live changes made to the CLI.
+
+1. Create a directory that will hold all the virtualenv packages and files. Note that you will only need to run this command once.
```
Comment (orenbm) [inline CONTRIBUTING.md:37]: virtualenv in `virtualenv`?

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -18,45 +18,163 @@ any additional information for contributing code to this project.
...
+
 ```
 $ python3 -m venv venv
 ```
 
-- Enable your terminal to use those files with this command:
+Windows:
+
+```
+py -m venv venv
```
Comment (orenbm) [inline CONTRIBUTING.md:48]: add `$`? or is this a linux thing?

File: CONTRIBUTING.md [hunk 8]
```diff
@@ -18,45 +18,163 @@ any additional information for contributing code to this project.
...
+1. As a process (no Python required)
+
+#### Running in a containerized environment
+
+To run specific tests, perform the following:
+
+1. Drop down into a test container
+
+```
+$ ./bin/test_integration -d
```
Comment (orenbm) [inline CONTRIBUTING.md:106]: what's `-d`? do we have a full-name flag?

File: CONTRIBUTING.md [hunk 9]
```diff
@@ -18,45 +18,163 @@ any additional information for contributing code to this project.
...
+1. Under the individual test add the following:
+
+```
+<name-of-test-function>.<identifier>=True
+
+## Example
+def my-integration-test-function()
+  ...
+
+my-integration-test-function.someidentifier=True
```
Comment (orenbm) [inline CONTRIBUTING.md:118]: what is `someidentifier`? if this is an example, give a real example.

File: CONTRIBUTING.md [hunk 10]
```diff
@@ -18,45 +18,163 @@ any additional information for contributing code to this project.
...
+root@123456:/opt/conjur-api-python3# nose2 -v -X --config integration_test.cfg -A 'someidentifier' $@
+```
+
+1. You should see that only that specific test is run. Every change made locally can be seen in the container so you
+do *not* need to rebuild before running these tests again.
+
+#### Running as a process
+
+We provide the option to run the integration tests as a process. This means integration tests can run in the same way
+a user would use our CLI; without Python installed on the machine, using the Conjur CLI exec.
```
Comment (orenbm) [inline CONTRIBUTING.md:136]: maybe replace `;` with `-`?

File: CONTRIBUTING.md [hunk 11]
```diff
@@ -18,45 +18,163 @@ any additional information for contributing code to this project.
...
+1. Pack the `integration_test_runner.py` using PyInstaller in the platform to run the executable.
+
+  To pack: `pyinstaller -F test/util/test_runners/integration_test_runner.py`. Note that you will need
+  to pack each runner in each platform that you want to run the tests.
+
+1. Pack the Conjur CLI using PyInstaller in the platform to run the executable
+
+  To pack: `pyinstaller -F ./pkg_bin/conjur`. Note that you will need to pack each runner in each platform
+  that you want to run the tests. Also note that _only_ for macOS, you should pack the CLI as a directory instead of a
+  single file (using -D instead of -F). This will allow the CLI tests to run quicker.
```
Comment (orenbm) [inline CONTRIBUTING.md:152]: `faster` instead of `quicker`?

File: CONTRIBUTING.md [hunk 12]
```diff
@@ -18,45 +18,163 @@ any additional information for contributing code to this project.
...
+
+1. Run the executable: `./integration_test_runner`, supplying the below required parameters via the command line.
+
+###### Required parameters
+
+`--invoke_cli_as_process`, required to run the CLI as a process.
```
Comment (orenbm) [inline CONTRIBUTING.md:158]: i am used to flags being separated by `-` instead of `_`. Is this intentional?

Comment (orenbm) [reply CONTRIBUTING.md:158]: i would change `,` to ` - ` or `:`

File: CONTRIBUTING.md [hunk 13]
```diff
@@ -18,45 +18,163 @@ any additional information for contributing code to this project.
...
+
+`--cli_to_test`, path to the packed CLI executable to test against.
+
+Parameters like --url, --account, --login, --password, will used to configure the CLI. These values are used
```
Comment (orenbm) [inline CONTRIBUTING.md:162]: will be used

File: CONTRIBUTING.md [hunk 14]
```diff
@@ -18,45 +18,163 @@ any additional information for contributing code to this project.
...
+before each test profile is run to configure the CLI and run the integration tests successfully.
+
+`--identifier`, the test method with this identifier will be run (`integration` by default).
+To run as a process, the identifier should be `test_with_process`.
+
+`--files_folder` path to test assets (policy files, etc). This folder is located under `/test/test_config` in the
+repo. Copy this executable into every OS you wish to run the CLI integration tests.
+
+
```
Comment (orenbm) [inline CONTRIBUTING.md:171]: remove this empty line

File: CONTRIBUTING.md [hunk 15]
```diff
@@ -18,45 +18,163 @@ any additional information for contributing code to this project.
...
+`--cli_to_test`, path to the packed CLI executable to test against.
+
+Parameters like --url, --account, --login, --password, will used to configure the CLI. These values are used
+before each test profile is run to configure the CLI and run the integration tests successfully.
+
+`--identifier`, the test method with this identifier will be run (`integration` by default).
+To run as a process, the identifier should be `test_with_process`.
+
+`--files_folder` path to test assets (policy files, etc). This folder is located under `/test/test_config` in the
```
Comment (orenbm) [inline CONTRIBUTING.md:168]: missing `,`

File: CONTRIBUTING.md [hunk 16]
```diff
@@ -18,45 +18,163 @@ any additional information for contributing code to this project.
...
+repo. Copy this executable into every OS you wish to run the CLI integration tests.
+
+
+###### Example
+```
+./integrations_tests_runner -i test_with_process -u https://conjur-server -a someaccount -l somelogin -p Myp@ssw0rd1!
```
Comment (orenbm) [inline CONTRIBUTING.md:174]: can we use full-name flags?

Comment (orenbm) [reply CONTRIBUTING.md:174]: also i would break this per param for readability:
```
./integrations_tests_runner \
  -i test_with_process \
  -u https://conjur-server \
  -a someaccount \
  -l somelogin \
  -p Myp@ssw0rd1! \
  -f /test \
  -c /dist/conjur \
  --invoke_cli_as_process
```

File: CONTRIBUTING.md [hunk 17]
```diff
@@ -72,11 +190,71 @@ See [here](guidelines/python-cli-ux-guidelines.md) for full UX guidelines to fol
...
+
+1. Review the git log and ensure the changelog contains all relevant recent changes with references to GitHub issues or PRs, if possible
+
+1. Review the changes since the last tag, and if the dependencies have changed revise the [NOTICES](NOTICES.txt) file to correctly capture the added dependencies and their licenses / copyrights
+
+1. Before creating a release, ensure that all documentation that needs to be written has been written by TW, approved by PO/Engineer, and pushed to the forward-facing documentation
+
+1. Commit these changes to the branch. Bump version to x.y.z is an acceptable commit message and open a PR for review
+
+Add a git tag
```
Comment (orenbm) [inline CONTRIBUTING.md:215]: i am guessing this should have `###`

File: CONTRIBUTING.md [hunk 18]
```diff
@@ -72,11 +190,71 @@ See [here](guidelines/python-cli-ux-guidelines.md) for full UX guidelines to fol
...
+
+1. Based on the unreleased content, determine the new version number and update the version in `version.py`
+
+1. Review the git log and ensure the changelog contains all relevant recent changes with references to GitHub issues or PRs, if possible
+
+1. Review the changes since the last tag, and if the dependencies have changed revise the [NOTICES](NOTICES.txt) file to correctly capture the added dependencies and their licenses / copyrights
+
+1. Before creating a release, ensure that all documentation that needs to be written has been written by TW, approved by PO/Engineer, and pushed to the forward-facing documentation
+
+1. Commit these changes to the branch. Bump version to x.y.z is an acceptable commit message and open a PR for review
```
Comment (orenbm) [inline CONTRIBUTING.md:213]: add `"` surrounding  `Bump version to x.y.z`

File: CONTRIBUTING.md [hunk 19]
```diff
@@ -72,11 +190,71 @@ See [here](guidelines/python-cli-ux-guidelines.md) for full UX guidelines to fol
...
+
+Add a git tag
+
+1. Once your changes have been reviewed and merged into master, tag the version using `git tag -s v0.1.1`. Note this requires you to be able to sign releases. Consult the [github documentation on signing commits](https://docs.github.com/en/free-pro-team@latest/github/authenticating-to-github/managing-commit-signature-verification) on how to set this up vx.y.z is an acceptable tag message
```
Comment (orenbm) [inline CONTRIBUTING.md:217]: missing period before `vx.y.z`. also add `"` surrounding `vx.y.z`

File: CONTRIBUTING.md [hunk 20]
```diff
@@ -72,11 +190,71 @@ See [here](guidelines/python-cli-ux-guidelines.md) for full UX guidelines to fol
...
+1. Push the tag: `git push vx.y.z` (or `git push origin vx.y.z` if you are working from your local machine)
```
Comment (orenbm) [inline CONTRIBUTING.md:218]: what do you mean by `if you are working from your local machine`? when am i not working from my local machine?

File: CONTRIBUTING.md [hunk 21]
```diff
@@ -72,11 +190,71 @@ See [here](guidelines/python-cli-ux-guidelines.md) for full UX guidelines to fol
...
+
+#### Linux
+
+1. Run `./bin/build_binary -l`. Once this is run, a `dist` folder will be created with the executable in it
+2. Once an executable has been created, run `tar -cvf conjur-cli-linux.tar conjur` to archive the file in a tar.gz format
+3. Add the archive file as an asset in the release
+
+#### macOS
+
+1. Run `./bin/build_binary -l`. Once this is run, a `dist` folder will be created with the executable in it
```
Comment (orenbm) [inline CONTRIBUTING.md:238]: use full-name flag if possible

File: CONTRIBUTING.md [hunk 22]
```diff
@@ -72,11 +190,71 @@ See [here](guidelines/python-cli-ux-guidelines.md) for full UX guidelines to fol
...
+2. Once an executable has been created, run `tar -cvf conjur-cli-darwin.tar conjur` to archive the file in a tar.gz format
```
Comment (orenbm) [inline CONTRIBUTING.md:239]: add ` around `tar.gz`

File: CONTRIBUTING.md [hunk 23]
```diff
@@ -72,11 +190,71 @@ See [here](guidelines/python-cli-ux-guidelines.md) for full UX guidelines to fol
...
+3. Add the archive file as an asset in the release
```
Comment (orenbm) [inline CONTRIBUTING.md:240]: how?

File: CONTRIBUTING.md (hunk 4 above)
Comment (sigalsax) [reply CONTRIBUTING.md:27]: egg contains all the source files for a Python compressed so basically ZIP files that you can import into the python program without unzipping

File: CONTRIBUTING.md (hunk 6 above)
Comment (sigalsax) [reply CONTRIBUTING.md:37]: are you suggesting I wrap virtual end in``?

File: CONTRIBUTING.md (hunk 7 above)
Comment (sigalsax) [reply CONTRIBUTING.md:48]: don't know but I can add it. I think it is a pretty generic symbol to mean terminal

File: CONTRIBUTING.md (hunk 5 above)
Comment (sigalsax) [reply CONTRIBUTING.md:51]: 1 OS to show? These steps also help in our release process so I prefer to leave them because they are a good reference

File: CONTRIBUTING.md (hunk 8 above)
Comment (sigalsax) [reply CONTRIBUTING.md:106]: -d is debug which brings up containers and throws us into the bash
```
if [ "$1" == "-d" ]; then
  DEBUG="true"
  shift
fi

...

if [[ "$DEBUG" == "true" ]]; then
  docker-compose exec test bash
  exit
fi
```

should I add a comment or allow for the `--debug` form as well?

File: CONTRIBUTING.md (hunk 9 above)
Comment (sigalsax) [reply CONTRIBUTING.md:118]: someidentifer is any identifer you choose. so someidentifier is and can be a real example

File: CONTRIBUTING.md (hunk 11 above)
Comment (sigalsax) [reply CONTRIBUTING.md:152]: same same

File: CONTRIBUTING.md (hunk 20 above)
Comment (sigalsax) [reply CONTRIBUTING.md:218]: took this from SP
https://github.com/cyberark/secrets-provider-for-k8s/blob/master/CONTRIBUTING.md#add-a-git-tag
which we probably took from Secretless

File: CONTRIBUTING.md (hunk 8 above)
Comment (sigalsax) [reply CONTRIBUTING.md:106]: will add `if [[ "$1" = "--debug" || "$1" = "-d" ]]; then`

File: CONTRIBUTING.md (hunk 12 above)
Comment (sigalsax) [reply CONTRIBUTING.md:158]: will change

File: CONTRIBUTING.md (hunk 15 above)
Comment (sigalsax) [reply CONTRIBUTING.md:168]: updated to `-`

File: CONTRIBUTING.md (hunk 16 above)
Comment (sigalsax) [reply CONTRIBUTING.md:174]: great idea for readability

File: CONTRIBUTING.md (hunk 21 above)
Comment (sigalsax) [reply CONTRIBUTING.md:238]:
```
BIN_BUILD_CMD="pyinstaller --onefile pkg_bin/conjur"

if [[ "$1" == "-l" ]]; then
  $BIN_BUILD_CMD
  exit 0
fi
```

so it packs the cli with this script otherwise spins up a container and packs the executable in that container. So I think -l is local.
 I will add a long form

File: CONTRIBUTING.md (hunk 23 above)
Comment (sigalsax) [reply CONTRIBUTING.md:240]: Create a release and just add the file.

Comment (sigalsax) [reply CONTRIBUTING.md:240]: Thought it is a known process?

File: CONTRIBUTING.md (hunk 6 above)
Comment (orenbm) [reply CONTRIBUTING.md]: yes

File: CONTRIBUTING.md (hunk 5 above)
Comment (orenbm) [reply CONTRIBUTING.md:51]: no. i meant using `1` for all steps and not use `2`

File: CONTRIBUTING.md (hunk 8 above)
Comment (orenbm) [reply CONTRIBUTING.md:106]: i think that's the best solution. in the example we should use the full-name

File: CONTRIBUTING.md (hunk 20 above)
Comment (orenbm) [reply CONTRIBUTING.md:225]: what does the girl-scout in you think of this?

File: CONTRIBUTING.md (hunk 23 above)
Comment (orenbm) [reply CONTRIBUTING.md:247]: ok

File: CONTRIBUTING.md (hunk 20 above)
Comment (sigalsax) [reply CONTRIBUTING.md:225]: i miss being a girlscout... 
Now that I think about it, I remember that for the SP releases I had to push to the origin to make it work which is probably why I added that note in the first place.

**Review by tzheleznyak — APPROVED** [review]
